Upstream staging sync PR

UnityAlex opened the pull request:
DRAFT for reviewing purposes just to see how bad this merge could be. :-)
<!--
Thank you for your Pull Request!

Here are a few things to think about (see below for more details). Please check each option after the PR is created.
-->

- Should this pull request have release notes?
  - [ ] Yes
  - [ ] No
- Do these changes need to be back ported?
  - [ ] Yes
  - [ ] No

Reviewers: please consider these questions as well! :heart:

<!-- Use this section if the pull request has release notes.
**Release notes**

Fixed case XXXXXX @username:
Mono: Your release notes go here.

Other options: Internal, Changed, Improved, Feature. 
-->

<!-- Use this section is the pull request should be back ported.
**Backports**

List the versions of Unity where this change should be back ported here.
-->

<!-- Use this section if the pull request requires other changes in the Unity repository.
**Unity repository changes**

List any Unity repository PRs.
-->

akoeplinger committed:
Build iOS SDK archives on Xcode 12.4

(cherry picked from commit eb4c3116ebf3ac44f72232fc9315fa4947c04b44)

SamMonoRT committed:
Add cred scan suppressions for approvals (#20956)

* Add cred scan suppressions for approvals

* Update RSACryptoServiceProviderTest.cs

* Update RSAManagedTest.cs

* Update Type3MessageTest.cs

* Update npdbtest.cs

* Update InMemorySymmetricSecurityKeyTest.cs

* Update EncryptedXmlTest.cs

* Update WSSecurityTokenSerializerTest.cs

* Update X509Certificate2Test.cs

SamMonoRT committed:
Update CredScanSuppressions based on approved exceptions (#20955)

Adding certificate files for approved exceptions.

Commit:
Added supression for cred scan. (#20963)

fanyang-mono committed:
Suppress CredScan checking (#20961)

* Suppress CredScan checking

* Add justification

* One more file to suppress

Commit:
Added credscan supressions. (#20964)

fanyang-mono committed:
Suppress CredScan checking (#20965)

akoeplinger committed:
Run only one concurrent sdks archive and osx pgk build on main in CI

fanyang-mono committed:
Fix CredScan suppression locations (#20966)

thaystg committed:
[debugger] Removing call to jit_done to help debug multithread (#20970)

* [debugger] Removing call to jit_done to help debug multithread

It was causing a side effect on Unity, I will study it again while implement multithread on icordebug.

* Update mini-runtime.c

* Update dtest.cs

Commit:
[Credscan] Added supression comment in Web.Config. (#20967)

imhameed committed:
CredScan: add suppressions for SharpZipLib (#20974)

akoeplinger committed:
Bump corefx to get credscan fixes

MaximLipnin committed:
Fix the System.String.Replace throwing NotImplementedException (#20960)

https://github.com/mono/mono/issues/20948

aik-jahoda committed:
Suppress FTP anonymous credentials (#20977)

akoeplinger committed:
Bump corefx to get latest credscan fixes

akoeplinger committed:
Bump msbuild, roslyn and nuget (#20979)

Bump to MSBuild vs16.9, Roslyn 3.9.0 and NuGet 5.9.0

akoeplinger committed:
Disable JSC runs on WASM since it crashes

See https://github.com/mono/mono/issues/20982

BrzVlad committed:
Fix r4_conv_to_i (#20985)

* [mini] Add missing opcodes for r4 to native int conversions

* [interp] Add handling for conv.i from r4

akoeplinger committed:
[ci] Use adoptopenjdk instead of zulu-8 in sdks-archive.groovy

rootwyrm committed:
Update to consume jemalloc upstream releases (#20988)

This updates the build to consume upstream jemalloc releases starting with 5.2.1 (Aug 2019) while preserving the public and private prefixing. This passed testing locally without issue, but definitely could use a once over by a more familiar set of eyes. I tested with both `--with-jemalloc` and `--use-jemalloc-always` successfully on Ubuntu, Alpine, and FreeBSD.

akoeplinger committed:
Bump fsharp.py to get upstream nuget feed cleanup

myget.org feeds were disabled recently.

spaarmann committed:
Fix setting or getting pointer fields with reflection (#20891)

Setting a pointer via reflection incorrectly added an extra layer of
indirection, and getting a pointer via reflection incorrectly
dereferenced the value once.

As a result, setting and then getting only via reflection worked
correctly, but doing either step by directly accessing the field and the
other via reflection was broken.

Also adjusts `FieldInfoTest` to test this, for both static and instance
fields (the set path especially is different for those.)

Fixes #20872

This change is released under the MIT license.

akoeplinger committed:
Remove i386 slice from Microsoft.CodeAnalysis.CSharp.dll.dylib

It runs into issues with not sufficient headerpad which breaks notarization, temporarily remove the i386 slice.

nealef committed:
Import s390x fixes from dotnet runtime (#20991)

Recent fixes from dotnet/runtime for s390x
- Tailcall 
- Array construction
- Unwinding

Endle committed:
Fix implicit conversion warning (#20861)

This change is released under the MIT license.

Signed-off-by: Zhenbo Li litimetal@gmail.com

Calinou committed:
Fix console warning spam in threadpool-io.c (#20980)

This is needed to fix
https://github.com/godotengine/godot/issues/21895.

Taken from Unity's Mono fork: https://github.com/Unity-Technologies/mono/commit/37dc7aae643fa1d86f183044379352eac074d96e

Co-authored-by: Jonathan Chambers <joncham@gmail.com>

jaykrell committed:
mono_breakpoint_clean_code always returns TRUE. (#16603)

UnityAlex committed:
Correcting proclib function that did not adhere to filemap wrappers (#20989)

During our upgrade of the mono runtime we ran into a situation where `mono_pe_file_map` was being called as part of opening an image now on android. This function did not currently adhere to the filemap wrapper functions so the difference in data types being used by Unity caused crashes.

BrzVlad committed:
[arm64] Fix wrong marshalling in gsharedvt transition (#21004)

Transitioning from ArgVtypeByRefOnStack to ArgVtypeByRef requires no marshalling. The reference ends up being saved on stack the same way and the stack slot just needs to be copied (in mono_arm_start_gsharedvt_call).

sancheolz committed:
[WinForms] Fix #19755: Make ToolStripMenuItem constructor handle shortcut parameter (#19896)

sancheolz committed:
[WinForms] Fixed dependence of the countdown in the System.Windows.Forms.Timer on the system time. (#20596)

If you change the system time back a minute, then the timer will work only after a minute, even if you set its timeout to 500 milliseconds.
This fix uses a system-independent timing with Stopwatch.GetTimestamp ()

sancheolz committed:
The OnPropertyValueChanged event should (#20618)

tell the event handler about the old value of the property being changed using the OldValue field of the PropertyValueChanghedEventArgs class, but actually reports the new one by reading it from the changed Value. Therefore, we save the old value before the change that occurs in SetValueCore.

nvoronchev committed:
[WinForms] DataGridView: Update editing control position on scroll (#20675)

This pull request fixes the following issue. Let's go to edit mode of some cell from a `DataGridView` control. Then appeared text box will keep its locations if one scroll the grid.

nvoronchev committed:
[WinForms] Form: Set `WM_CLASS` while change `Text` (#20687)

The _Application Menu Button_ at the _Gnome Shell_ displays wrong application name -- it contains just a word _Unknown_. 

It seems this issue stems from the function `shell_app_get_name` in the https://gitlab.gnome.org/GNOME/gnome-shell/-/blob/gnome-3-38/src/shell-app.c#L256. Hence one need to set `WM_CLASS` to fix app name.

nvoronchev committed:
[WinForms] GridEnty: Invoke type converter to display entry value properly (#20766)

Type Convertor isn't invoked for `GridEntry` value at the present time. Current PR is aimed to fix this.

sancheolz committed:
When the OnClick event occurs on a Button mouse click, MouseEventArgs should be passed to it instead of the usual EventArgs. Then in the event handler you can see which key is pressed and the coordinates of the mouse cursor. Now the behavior is the same as in .net (#20871)

When the OnClick event occurs on a Button mouse click, MouseEventArgs should be passed to it instead of the usual EventArgs. Then in the event handler you can see which key is pressed and the coordinates of the mouse cursor. Now the behavior is the same as in .net

sancheolz committed:
[WinForms] When the ListView has VirtualMode=true, calling the Clear method for Items should not throw an exception. This is how .NET behaves (#20877)

nvoronchev committed:
[WinForms] ContextMenuStrip: Close toplevel menu if some click somewhere outside of it (#20953)

nvoronchev committed:
[WinForms] Add `ContextMenu` and `ContextMenuStrip` events for UIA API needs (#20957)

* [WinForms] ContextMenuStrip: Add events for UIA API needs

* [WinForms] ContextMenu: Add events for UIA API needs

akoeplinger committed:
Bump api-snapshot submodule

akoeplinger committed:
AzDO: Bump default timeout to 180mins

lambdageek committed:
[System.Native] Handle ReadDir EINTR (#21016)

* [System.Native] Handle ReadDir EINTR

Handle EINTR for opendir/readdir/closedir.

Because of how thread aborts work, we must loop in managed.
A thread abort will break syscalls and return EINTR, but mono will not raise a
ThreadAbortException until a pinvoke returns.

(Despite the documentation, opendir/readdir/closedir sometimes return EINTR on
macOS Big Sur.)

Related to https://github.com/mono/mono/issues/20799

* bump corefx

* [System] on unix, include Interop.ReadDir.Mono.cs

on mac it's in corlib, on unix and android it's in System, for some reason

* Add newline

Co-authored-by: Alexander Köplinger <alex.koeplinger@outlook.com>

lambdageek committed:
[MonoIO] Wrap calls to open() in EINTR handling (#21041)

Related to https://github.com/mono/mono/issues/21040 and
https://github.com/dotnet/runtime/issues/48663

On MacOS Big Sur open() is much more likely to throw EINTR - and moreover if
the thread receives a signal while doing an open() even if the signal handler
is marked with SA_RESTART, the open call appears not to restart and to fail
anyway.

So wrap all the calls to open() in retry loops.

Commit:
Azure Pipelines Windows CI (#20879)

* Disable EventLog tests on Windows if we're not running as admin

* Disable a socket test that doesn't work on Win10

* Fix MetadataResolverTest on Windows in AzDO

There's something running on localhost already that makes the test fail, use a different hostname instead.

* Force ASCII encoding for console-output.cs

* Use different way to compare output of test-console-output

* Fix issue with symlink on PublishTestResults

* Use Unicode encoding in Win32EventLog p/invokes, fixes eventlog tests on Win10

Co-authored-by: Alexander Köplinger <alex.koeplinger@outlook.com>

madewokherd committed:
Remove !sig->is_inflated assert from icall wrapper generator. (#21046)

If the method or type containing the icall instruction has a
generic parameter, the resulting signature is "inflated" even if
the generic parameter is not used in the signature. This breaks
some methods in SlimDX, and probably other C++/CLI code.

Co-authored-by: Esme Povirk <esme@codeweavers.com>

akoeplinger committed:
sdks: Disable parallelism for mxe llvmwin64 build

For some reason it recently fails with weird race conditions, let's see if that helps.

akoeplinger committed:
sdks: Use https for xamjenkinsartifacts llvm archive download

Accessing the storage account fails now when using plain http.

monojenkins committed:
[mono] Fix leak in assembly-specific dllmap lookups (#21048)

Should address https://github.com/mono/mono/issues/21034 when mirrored

The leaked memory was the duplicated strings after the first assembly-specific lookup. If that lookup failed, they were overwritten with the global lookup and never freed.

Co-authored-by: CoffeeFlux <CoffeeFlux@users.noreply.github.com>

akoeplinger committed:
Fix tarball build

nealef committed:
Add thunking to s390x for calls that will be patched (#21052)

* Import s390x fixes from dotnet runtime

* Add thunking to s390x for calls that will be patched

A problem arose when running with 8 cores and a call was being patched via the mono_arch_patch_callsite API. Because the value being patched may not be on a boundary that the architectures as being "block concurrent" (e.g. pointers need to be aligned on 8 byte boundaries). The opshot of which is that the value may be being updat
ed while another thread is fetching it and if not on a boundary the value could be in an intermediate state resulting in a branch to the wrong place.

This patchset implements thunking so that when a call is emitted the call sequence changes from:

        iihf %rx,<hiword of target address>
        iilf %rx,<loword of itarget address>
        bras %r14,%rx

to:

        lgrl %rx,thunk
        bras %r14,%rx

thunk:  .quad <target address>

Similarly, tailcalls change from:

        iihf %r1,<hiword of target address>
        iilf %r1,<loword of itarget address>
        j    %r1

to:

        lgrl %r1,thunk
        br   %r1

In addition, the file support-s390x.h is a hangover and can now be rolled into mini-s390x.h.

* mono/mini/Makefile.am.in:
  - Remove support-s390x.h
* mono/mini/exceptions-s390x.c:
  - Eliminate a couple of compiler cast warnings
* mono/mini/mini-s390x.c:
  - Implement thunking
* mono/mini/mini-s390x.h:
  - Incorporate support-s390x.h
* mono/mini/support-s390x.h:
  - Delete file
* mono/mini/tramp-s390x.c:
  - Thunking changes

* Remove commented out code

thaystg committed:
[debugger][interpreter] Generate seq_points even when doesn't have debug information (#21058)

https://github.com/dotnet/runtime/pull/52632/

mrward committed:
[MSBuild] Update to vs16.10 branch (#21083)

jbcoe committed:
Fix for null pointer bug found by UBSAN (#21070)

Fix for null-pointer bug found with Clang's undefined-behavior-sanitizer using `-fsanitize=null`

https://clang.llvm.org/docs/UndefinedBehaviorSanitizer.html

akoeplinger committed:
[ci] Disable C++ AzDO jobs

We won't switch to C++ in the mono/mono repo.

bholmes committed:
Fixing the GetInterfaceMap tests for mono in the runtime repo. (#21094)

Addressing 4 issues for GetInterfaceMap and default interface methods

Only methods marked as virtual on on interface should be added to the
interface map. (no static or instance)
If the found target method is ambiguous (a diamond) the target is
null.
If the found target method's class in an interface, then the target
class is the interface class, else it is the class of the RuntimeType
(aka this)
If the found target method is abstract (reabstraction) then the
target is null.

https://github.com/dotnet/runtime/blob/6072e4d/src/tests/reflection/DefaultInterfaceMethods/GetInterfaceMapConsumer.cs

adamsitnik committed:
[mono][wasm] add new signatures for pread|pwrite, preadv|pwritev, posix_fallocate (#21098)

bholmes committed:
Increase the buffer size for ping reply (#21069)

Avoid Windows socket error code WSAEMSGSIZE when receiving the ping
reply for larger sent payloads.

bluerise committed:
Add support for OpenBSD/arm64 (#21015)

Provide the UCONTEXT macros, which for OpenBSD simply map to sigcontext.

akoeplinger committed:
Bump boringssl submodule

thaystg committed:
[debugger] Fix assertion when trying to debug an async method in a valuetype (#21100)

* Fix * Assertion at C:\build\output\Unity-Technologies\mono\mono\mini\debugger-engine.c:1657, condition `is_ok (error)' not met, function:set_set_notification_for_wait_completion_flag, Could not execute the method because the containing type 'System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1[T_REF]', is not fully instantiated. assembly:<unknown assembly> type:<unknown type> member:(null)

* Adding test case.

* Remove unrelated changes.

* Fix test on interpreter.

sancheolz committed:
[WinForms] MenuStrip: Close toplevel menu if some click somewhere outside of it (#21014)

Fix regression introduced in # 20953
If we do not use this typecasting, then NRE will occur in line 910 for MenuStrip with submenu embedded in the form.

sancheolz committed:
[WinForms] Fix changing reference to PrintController in preview in PrintPreviewControl (#21010)

If a reference to PrintDocument is passed to PrintPreviewControl, and then this control is displayed on the form,
then original PrintController of the PrintDocument will be replaced by previewController and if you call print for PrintController,
then the print page will not be sent to printer.

gardenappl committed:
Allow manipulating console window size in Xterm (#20886)

Added implementations for `Console.SetWindowSize(int, int)`, `Console.SetBufferSize(int, int)` and associated properties. This is [based on Xterm control sequences](https://invisible-island.net/xterm/ctlseqs/ctlseqs.html#h3-Miscellaneous) which are supported by a lot of Linux terminals.

The previous API was very hit or miss - e.g. the method `Console.SetWindowSize(int, int)` just does absolutely nothing, while the setter for `Console.WindowWidth` always throws an exception.

The new implementation sends a control sequence if `TERM` starts with "xterm", otherwise throws an exception.

akoeplinger committed:
Bump api-snapshot submodule

akoeplinger committed:
Bump api-snapshot submodule

fanyang-mono committed:
Fix memory leak during data registration (#21107)

lambdageek committed:
Don't return null target methods in GetInterfaceMap for abstract classes, or classes using DIMs (#21112)

* Don't return null target methods in GetInterfaceMap for abstract classes, or classes using DIMs

If the interface method is reabstracted, and either the found implementation
method is abstract, or the found implementation method is from another
DIM (meaning neither klass nor any of its ancestor classes implemented the
method), then say the target method is null. Otherwise return the found
implementation method, even if it is abstract.

Corresponding dotnet/runtime PR is https://github.com/dotnet/runtime/pull/53972

Fixes https://github.com/dotnet/runtime/issues/53933

* Add method flags accessors; cleanup interface map code

Ported from https://github.com/dotnet/runtime/pull/54271

lambdageek committed:
Start a dedicated thread for MERP crash reporting (#21096)

Related to https://github.com/mono/mono/issues/21009 

There are two scenarios:
1.  When someone force quits Mono (or just runs `kill -TERM <pid>`), the process can receive the signal on any thread,
2. or - if a thread in the process crashes, but the thread is not attached to the runtime, Mono's signal handlers still run.

The crash reporter assumes that the crashing thread is either attached to the runtime, or at least `mono_thread_info_current` or the JIT TLS data are set for the thread.  If the thread is truly foreign and it never interacted with Mono, and it crashes, both of those assumptions are false, but Mono's crash reporter signal handlers still run.

The solution from this PR is: if crash reporting is enabled, start a dedicated thread at process startup that is a "crash report leader" - when a crash happens, the crashing thread (the crash originator) wakes the leader, and the leader collects the crash report.  The crash originator does not do any work that requires being attached to the runtime or to the JIT such as iterating over thread IDs or stack walking.

---

* Add a standalone test of crash on a foreign thread

* extra async printfs

* Sketch out crash leader implementation

   At process startup, start a separate thread that is attached to the runtime and can collect crash reports.  Crashing threads will wake it and wait for it to collect the crash reports

* Start moving responsibilities to the crash leader

* Watch for nil jit_tls data if crash is on a foreign thread

* Start adding a state machine for the crash leader

   We need to coordinate the originator and the leader in a few places.

   The leader needs to pause to after collecting the thread ids before suspending the non-originator threads, and again while the originator is dumping its own stack.

   The originator needs to wait for the leader to collect the thread IDs and to tell it its assigned slot. Then it tells the leader to suspend the others, dumps its own memory, then tell the leader to dump the whole crash report and wait for it to reply when it's done.

* Move remaining summarizer work to the summarizer leader thread

* change some debug printfs

* Successfully summarize a foreign thread crash

* Add test to merp-crash-test.cs

* delete standalone test

* if the crash leader is the originator, collect the report synchronously

   either because the crash leader crashed, or because the process got a SIGTERM and it arrived on the crash leader thread

* fixup comments

* turn off crash leader printf debugging

* Don't create leader thread if crash reporting is disabled

* Fix test on win32

* Update mono/tests/libtest.c

* fixup logging and comments around managed stack collection

* Disable the merp dlopen crasher

* Remove crash leader state machine

   It's straightline code with two early exits. State machine is overkill

* remove remnants of leader state machine

dseshadri committed:
Fix for heap_use_after_free flagged by sanitizer (#21120)

This change is released under the MIT license.

akoeplinger committed:
Remove CoffeeFlux from CODEOWNERS

https://github.com/dotnet/runtime/commit/3b3099f00add243f7ee453e7fd6975a00a9871ed

DzmitryKN committed:
optimize adding unwind info entries by hashtable (#21128)

* optimize adding unwind info entries by hashtable

* cosmetic

* fix code styling

monojenkins committed:
[mono] Fix race during mono_image_storage_open (#21142)

* The mono_refcount_inc call in mono_image_storage_trypublish or
  mono_image_storage_tryaddref may abort when racing against a
  mono_image_storage_dtor that already decremented the refcount.

This race triggered in some cases when building aspnetcore using a Mono-based dotnet host SDK.   The problem is that `mono_image_storage_close` runs outside the `mono_images_storage_lock` (and this may be unavoidable due to lock ordering concerns).  Therefore, we can have a refcount that was already decremented to zero, but before `mono_image_storage_dtor` finishes removing the object from `images_storage_hash`, a parallel `mono_image_storage_trypublish` may have retrieved it from there.  In that case, the `mono_refcount_inc` call will abort.

Fixed by detecting that case via `mono_refcount_tryinc` instead, and simply treating the object as if it had already been removed.  It will in time actually get removed, either by the parallel `mono_image_storage_dtor`, or else by the `g_hash_table_insert` in `mono_image_storage_trypublish` (which will safely replace it with the new object, and `mono_image_storage_dtor` will then detect that and skip removal).

Co-authored-by: uweigand <uweigand@users.noreply.github.com>

monojenkins committed:
[sgen] Transition to GC Unsafe in mono_gc_wait_for_bridge_processing (#21157)

Mark it as an external only API.  In the runtime, use `mono_gc_wait_for_bridge_processing_internal`

Co-authored-by: lambdageek <lambdageek@users.noreply.github.com>

nealef committed:
Enable Float32, fix linears optimization, and rework localloc (#21161)

* - Enable full float32 support
  - Add instructions for FP operations: abs[f]/ceil[f]/floor[f]/round[f]/sqrt[f]/trunc[f]
  - Enable these instructions in mono_arch_emit_inst_for_method()
  - Handle return of float32 results
  - Correct rounding mode for OP_LCONV_TO_R_UN processing
- Enable MONO_OPT_LINEAR optimizations
  - Correct prolog processing of structure returned variable
- Rework OP_LOCALLOC for cases where alloc size > 4k
- Add OP_POPCNTxx support
- Minor typo (missing tab)

* For 17 years I've believed that structures of size 1, 2, 4, or 8 bytes could be returned in a register. TIL this is not the case.

* While we don't return structures in the register we can still optimize so that we
keep the address of the structure in a register so we need to properly load it in
the prolog. Similarly, in the epilog we need to set the return register with that
value.

I would have thought that the SETRET processing would have taken care of this which
it does when we put the address on the stack but for the case when that variable is
kept in a register we don't seem to automatically load the return register.

* My prevous comments wrt mono_arch_emit_epilog were incorrect. There is no need to explicitly
set the register here as it will be done. My interpretation of my test case was incorrect.

nealef committed:
Missed this float32 instruction: OP_RMOVE (#21167)

* Missed this float32 instruction: OP_RMOVE

* Additional float32 fixes
- rmove specification in cpu-s390x.md
- OP_RMOVE implementation in mini-s390x.c
- Correct float instruction lengths
- Disable debug code

monojenkins committed:
[mono] Fix an uninitialized memory access in the ABCREM pass. (#21175)

Detected by valgrind.

Co-authored-by: vargaz <vargaz@users.noreply.github.com>

lambdageek committed:
[mini] Add GC Unsafe transitions in mono_pmip (#21185)

Add a new mono_pmip_u that doesn't do the transition.

The intent is that Mono developers in the debugger can call still call
mono_pmip_u if the thread state machine is in a broken state.

UnityAlex committed:
Merge remote-tracking branch 'upstream/main' into upstream-staging

UnityAlex committed:
fixing merge error to get things building again

unity-cla-assistant commented:
[![CLA assistant check](https://cla-assistant.cds.internal.unity3d.com/pull/badge/not_signed)](https://cla-assistant.cds.internal.unity3d.com/Unity-Technologies/mono?pullRequest=1479) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.cds.internal.unity3d.com/Unity-Technologies/mono?pullRequest=1479) before we can accept your contribution.<br/>**0** out of **33** committers have signed the CLA.<br/><br/>:x: naricc<br/>:x: rootwyrm<br/>:x: SamMonoRT<br/>:x: imhameed<br/>:x: spaarmann<br/>:x: akoeplinger<br/>:x: aik-jahoda<br/>:x: fanyang-mono<br/>:x: MaximLipnin<br/>:x: thaystg<br/>:x: jaykrell<br/>:x: Endle<br/>:x: BrzVlad<br/>:x: madewokherd<br/>:x: sancheolz<br/>:x: lambdageek<br/>:x: bluerise<br/>:x: nealef<br/>:x: monojenkins<br/>:x: Calinou<br/>:x: jbcoe<br/>:x: AtariDreams<br/>:x: gardenappl<br/>:x: mcdon-ron<br/>:x: DzmitryKN<br/>:x: csanchezdll<br/>:x: NVoronchev<br/>:x: mrward<br/>:x: directhex<br/>:x: adamsitnik<br/>:x: dseshadri<br/>:x: rdavisau<br/>:x: EhsanKia<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.cds.internal.unity3d.com/check/Unity-Technologies/mono?pullRequest=1479) it.</sub>

rdavisau committed:
[Mono.Profiler.Aot] Write true string wire length (#21191)

In some cases, it seems like Mono.Profiler.Aot `ProfileWriter` can produce custom profiles that are unable to be read by `ProfileReader` (or `aot-compiler.c`). There are a few reports [here](https://github.com/xamarin/xamarin-android/issues/4602#issuecomment-618201419).

I investigated and found a cause may be the writing of incorrect string length markers in situations where `String.Length` is less than the encoded byte length. In that kind of case, the reader falls out of sync with the record structure and quickly fails. [Here is an example dependency](https://www.fuget.org/packages/AuroraControls.Core/1.2020.520.2/lib/netstandard2.0/Aurora.dll) with members that can produce the issue (I guess the odd characters are caused by obfuscation).

![image](https://user-images.githubusercontent.com/7392704/129899400-086a6e8a-373c-4601-837a-5255e7f93d82.png)

I don't think the change requires a profile format version bump, as the format itself is unchanged.

bholmes committed:
Adding null check to avoid abort when invalid IL is encountered (#21194)

When a callvirt instruction is encountered on a static method, there is
a hard abort/crash. This commit avoids the problem.

akoeplinger committed:
Add backport GitHub Action

LumioseSil committed:
Remove dead code in blts-error (#21155)

This return statement is dead code and can’t be reached. Remove it.

bholmes committed:
Update portable executable and import table validation (#21109)

* Update portable executable and import table validation

Update the SectionAlignment and FileAlignment validate to match the
ms docs...

https://docs.microsoft.com/en-us/windows/win32/debug/pe-format#optional-header-windows-specific-fields-image-only

"SectionAlignment
The alignment (in bytes) of sections when they are loaded into memory
It must be greater than or equal to FileAlignment. The default is the
page size for the architecture."

"FileAlignment
The alignment factor (in bytes) that is used to align the raw data of
sections in the image file. The value should be a power of 2 between 512
and 64 K, inclusive. The default is 512. If the SectionAlignment is less
than the architecture's page size, then FileAlignment must match
SectionAlignment."

For the import table use a case insensitive compare for the mscoree.dll
file name.

* Fixing a error message typo...  

"Aligmnent" to "Alignment"

mcdon-ron committed:
Allow Case Insensitive OptionSet (#21192)

* Added constructors to OptionSet to allow for Case Insensitive command line arguments. Will be Case Sensitive by default, and can pass StringComparer.OrdinalIgnoreCase to use Case Insensitive which is passed to the base KeyedCollection<string, Option>. Added related unit tests in CaseSensitivityTests.cs.

* Fix code style

* Update Mono.Options xml docs

* Bump api-snapshot submodule

Co-authored-by: Alexander Köplinger <alex.koeplinger@outlook.com>

EhsanKia committed:
Add missing comma leading to unwanted str concat (#21197)

The missing comma leads to "SET_KEEPALIVE" and "GET_TYPES_FOR_SOURCE_FILES" being merged, breaking both entries. This is very likely unwanted and a bug.

csanchezdll committed:
Fix pseudo-ttys for kernel versions >= 5.13 (#21203) (#21204)

monojenkins committed:
[mini] Don't add unbox tramopline on generic DIM calls (#21208)

Don't unbox a valuetype `this` if the generic method is a DIM

Fixes https://github.com/dotnet/runtime/issues/58394

Co-authored-by: lambdageek <lambdageek@users.noreply.github.com>

madewokherd committed:
Fix ordinal imports on Windows. (#21123)

Co-authored-by: Esme Povirk <esme@codeweavers.com>

UnityAlex committed:
Merge remote-tracking branch 'upstream/main' into upstream-staging
